web/src/ui_util.ts

```diff
@@ -44,6 +44,62 @@ export function change_katex_to_raw_latex($element: JQuery): void {
     });
 }
 
+export function potentially_collapse_quotes($element: JQuery): void {
+    const $children = $element.children();
+    const collapsible_status = [...$children].map((element) => is_collapsible($(element)));
+
+    if (collapsible_status.every(Boolean)) {
+        // If every element is collapsible, we don't collapse any of them.
+        return;
+    }
+
+    collapse_quotes($children, collapsible_status);
+}
+
+function collapse_quotes($elements: JQuery, collapsible_status: boolean[]): void {
+    $elements.each((index, element) => {
+        if (collapsible_status[index]) {
+            if (index > 0 && collapsible_status[index - 1]) {
+                // If the previous element was also collapsible, remove this element
+                // to have a single collapsed block instead of multiple in a row.
+                $(element).remove();
+            } else {
+                // Else, collapse this element.
+                $(element).replaceWith($("<p>").text("[...]"));
+            }
+        }
+    });
+}
+
+function is_collapsible($element: JQuery): boolean {
+    return $element.is("blockquote") || is_user_said_paragraph($element);
+}
```

Comment:
Sort these functions in dependency order, so `is_user_said_paragraph` first, this second, etc.